napari/components/layerlist.py

```diff
@@ -1,36 +1,50 @@
 from typing import Optional, List
 from ..layers import Layer
 from ..utils.naming import inc_name_count
-from ..utils.events.containers import TypedEventedList
+from ..utils.events import TypedEventedList
 
 
 class LayerList(TypedEventedList):
     """List-like layer collection with built-in reordering and callback hooks.
 
     Parameters
     ----------
-    data : iterable
+    iterable : iterable
         Iterable of napari.layer.Layer
+
+    Attributes
+    ----------
+    events : vispy.util.event.EmitterGroup
+        Event hooks:
+            * added(item, index): whenever an item is added
+            * removed(item): whenever an item is removed
+            * reordered(): whenever the list is reordered
     """
 
     def __init__(self, data=()):
         super().__init__(
-            basetype=Layer, data=data, lookup={str: lambda x: x.name},
-        )
-
-        self.events.inserted.connect(self._add)
-
-    def __newlike__(self, iterable):
-        return TypedEventedList(
-            basetype=self._basetype, data=iterable, lookup=self._lookup
+            data=data, basetype=Layer, lookup={str: lambda e: e.name},
         )
 
-    def _add(self, event):
-        """When a layer is added, set its name."""
-        layer = event.value
-        layer.name = self._coerce_name(layer.name, layer)
-        layer.events.name.connect(lambda e: self._update_name(e))
-        self.unselect_all(ignore=layer)
+    def _prepare_to_add_layer(self, layer: Layer):
+        if isinstance(layer, Layer):
+            layer.name = self._coerce_name(layer.name, layer)
+            layer.events.name.connect(lambda e: self._update_name(e))
+
+    def insert(self, index: int, value):
+        self._prepare_to_add_layer(value)
+        super().insert(index, value)
+        self.unselect_all(ignore=value)
+
+    def __setitem__(self, key, value):
+        self._prepare_to_add_layer(value)
+        super().__setitem__(key, value)
+        # FIXME: the behavior of unselecting everything but the last-added
```

Comment:
I wouldn't call the unselection behavior error prone - it worked fine until what has been a major change in how events are emitted. I think I still want the behavior that adding a new layer leads to only it being selected, but overwriting/changing an existing one doesn't need to change selection - it should though do the name checking which it wasn't before though, and now it

Reply:
sorry, that was poor wording on my part.  I agree with you, the "behavior" is fine.   By "error" prone i meant that you need to change the list "just the right way" for it to work.  So perhaps i mean it's a bit brittle.  The intention here isn't to have a major change in the how the events are emitted... just to make sure that they aren't emitted selectively, like when you use `pop()` or `insert()` but not when you use `clear()` or slice style insertion ... We've gone through effort to make the layer list behavior like a "normal" list, so we have to expect/prepare for users using _any_ of the valid list methods (_or_ we have to make them specifically inaccessible)